vault/resource_ldap_auth_backend.go

```diff
@@ -0,0 +1,296 @@
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"groupdn": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"groupattr": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"path": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Default:  "ldap",
+				StateFunc: func(v interface{}) string {
+					return strings.Trim(v.(string), "/")
+				},
+			},
+		},
+	}
+}
+
+func ldapAuthBackendConfigPath(path string) string {
+	return "auth/" + strings.Trim(path, "/") + "/config"
+}
+
+func ldapAuthBackendWrite(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	authType := ldapAuthType
+	path := d.Get("path").(string)
+	desc := d.Get("description").(string)
+
+	log.Printf("[DEBUG] Enabling LDAP auth backend %q", path)
+	err := client.Sys().EnableAuth(path, authType, desc)
+	if err != nil {
+		return fmt.Errorf("Error enabling LDAP auth backend %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Enabled LDAP auth backend %q", path)
+
+	d.SetId(path)
+
+	return ldapAuthBackendUpdate(d, meta)
+}
+
+func ldapAuthBackendUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	path := ldapAuthBackendConfigPath(d.Id())
+	data := map[string]interface{}{}
+
+	if v, ok := d.GetOk("url"); ok {
+		data["url"] = v.(string)
+	}
+
+	if v, ok := d.GetOkExists("starttls"); ok {
+		data["starttls"] = v.(bool)
+	}
+
+	if v, ok := d.GetOk("tls_min_version"); ok {
+		data["tls_min_version"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("tls_max_version"); ok {
+		data["tls_max_version"] = v.(string)
+	}
+
+	if v, ok := d.GetOkExists("insecure_tls"); ok {
+		data["insecure_tls"] = v.(bool)
+	}
+
+	if v, ok := d.GetOk("certificate"); ok {
+		data["certificate"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("binddn"); ok {
+		data["binddn"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("bindpass"); ok {
+		data["bindpass"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("userdn"); ok {
+		data["userdn"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("userattr"); ok {
+		data["userattr"] = v.(string)
+	}
+
+	if v, ok := d.GetOkExists("discoverdn"); ok {
+		data["discoverdn"] = v.(bool)
+	}
+
+	if v, ok := d.GetOkExists("deny_null_bind"); ok {
+		data["deny_null_bind"] = v.(bool)
+	}
+
+	if v, ok := d.GetOk("upndomain"); ok {
+		data["upndomain"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("groupfilter"); ok {
+		data["groupfilter"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("groupdn"); ok {
+		data["groupdn"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("groupattr"); ok {
+		data["groupattr"] = v.(string)
+	}
+
+	log.Printf("[DEBUG] Writing LDAP config %q", path)
+	_, err := client.Logical().Write(path, data)
+
+	if err != nil {
+		d.SetId("")
```

Comment:
I'm not sure if this is the "correct" way of handling this.  The idea here is that the ID gets set initially inside the Create function (ldapAuthBackendWrite).   This is the Update function (ldapAuthBackendUpdate) where the ID should already have been set.  

This code basically says "if we failed to update this resource inside Terraform, we should remove it from the state entirely because its probably not there any longer".

There were some other resources with this behaviour at the same time which I copied, but I'm not sure its the correct way to handle an error here.